source/slang/slang-ir-legalize-image-subscript.cpp

```diff
@@ -0,0 +1,196 @@
+                }
+                else
+                {
+                    newValue = legalizedStore;
+                    if (getIRVectorElementSize(imageElementType) != 4)
+                    {
+                        newValue = builder.emitVectorReshape(
+                            builder.getVectorType(
+                                vectorBaseType, builder.getIntValue(builder.getIntType(), 4)),
+                            newValue);
+                    }
+                }
+
+                storeParams[kValueParamIndex] = newValue;
+                auto imageStore = builder.emitImageStore(
+                    builder.getVoidType(),
+                    storeParams);
+                storeInst->replaceUsesWith(imageStore);
+                storeInst->removeAndDeallocate();
+                if (!imageSubscript->hasUses())
+                {
+                    imageSubscript->removeAndDeallocate();
+                }
+            }
+            break;
+        case kIROp_SwizzledStore:
+            {
+                auto swizzledStore = cast<IRSwizzledStore>(storeInst);
+                // Here we assume the imageElementType is already lowered into float4/uint4 types from any
+                // user-defined type.
+                assert(imageElementType->getOp() == kIROp_VectorType);
+                auto originalValue = builder.emitImageLoad(vector4Type, loadParams);
+                Array<IRInst*, 4> indices;
+                for (UInt i = 0; i < swizzledStore->getElementCount(); i++)
+                {
+                    indices.add(swizzledStore->getElementIndex(i));
+                }
+                auto newValue = builder.emitSwizzleSet(
+                    vector4Type,
+                    originalValue,
+                    swizzledStore->getSource(),
+                    swizzledStore->getElementCount(),
+                    indices.getBuffer());
+                storeParams[kValueParamIndex] = newValue;
+                auto imageStore = builder.emitImageStore(
+                    builder.getVoidType(),
+                    storeParams);
+                storeInst->replaceUsesWith(imageStore);
+                storeInst->removeAndDeallocate();
+                if (!imageSubscript->hasUses())
+                {
+                    imageSubscript->removeAndDeallocate();
+                }
+            }
+            break;
+        default:
+            break;
+        }
+    }
+    void legalizeImageSubscript(TargetRequest* target, IRModule* module, DiagnosticSink* sink)
+    {
+        IRBuilder builder(module);
+        for (auto globalInst : module->getModuleInst()->getChildren())
+        {
+            auto func = as<IRFunc>(globalInst);
+            if (!func)
+                continue;
+            for (auto block : func->getBlocks())
+            {
+                auto inst = block->getFirstInst();
```

Comment:
can this 
```
auto inst = block->getFirstInst();
IRInst* next;
for ( ; inst; inst = next)

```

be

```
for (auto inst : block->getChildren())
```

Reply:
This cannot be done since the order of `block->getChildren()` is not equivalent to the order of `next = inst->getNextInst();`

Reply:
Does the order matter here?

Reply:
If you are modifying the indications while iterating them, you should call 
for (auto inst: getModifiableChildren()).

All other iterate-while-modify loops are not correct due to automatic inst hoisting.

For example, if inside a function you have:

X = call ...
y = makeStruct X, 1
z = call f y

Now when you iterate the code and during the first iteration you replaces X to 2.

That makes the y inst hoistable to global scope after the modification. And now the next iteration you set inst to y, but 
y = makeStruct 2,1
is now at global scope so you will be processing the insts at global scope from iteration 2 and all the remaining insts in the function are not processed.

The only safe way to modify insts in a loop is to first copy all children insts into a list, and then work through that list. This is what getModifiableChildren does.